File: c++/merge_sort.cpp
/*



**** recursive merge sort ****

*/
#include <stdio.h>
#include<stdlib.h>
#include <iostream>

using namespace std;

void displayIt(int arr[],int length) {
    int i;

    for (i = 0; i < length; i++)
        printf("%d ",arr[i]);

    cout << "\n";
}

void merge1(int A[],int l,int mid,int h) {
    int i = l,j = mid + 1,k = l;
    int B[100];

    while (i <= mid && j <= h) {
        if (A[i] < A[j])
            B[k++] = A[i++];
        else
            B[k++] = A[j++];
    }

    // clean-up operations
    for (; i <= mid; i++)
        B[k++] = A[i];
    for (; j <= h; j++)
        B[k++] = A[j];

    for (i = l; i <= h; i++)
        A[i] = B[i];
}

void mergeSort(int A[],int l,int h) {
    int mid;

    if (l < h) {
        mid = (l + h) / 2;
        mergeSort(A,l,mid);
        mergeSort(A,mid + 1,h);
        merge1(A,l,mid,h);
    }
}

int main() {
    cout << "**** recursive merge sort ****\n";
    int arr[]={14,22,9,34,2,55,17,92,15,3,-3,0};
    int length = sizeof(arr) / sizeof(arr[0]);
    int i;

    cout << "before merge sort: " << endl;
    displayIt(arr,length);

    mergeSort(arr,0,11);

    cout << "after merge sort: \n";
    displayIt(arr,length);

    return 0;
}

File: java/MergeSort.java
/*




 **** merge sort ****

 */
import java.util.Arrays;

public class MergeSort {
  static void merge1(int[] A,int left,int middle,int right) {
    int[] leftTmpArray = new int[middle - left + 2];
    int[] rightTmpArray = new int[right - middle + 1];

    for (int i = 0; i <= middle - left; i++) {
      leftTmpArray[i] = A[left + i];
    }
    for (int i = 0; i < right - middle; i++) {
      rightTmpArray[i] = A[middle + 1 + i];
    }

    leftTmpArray[middle - left + 1] = Integer.MAX_VALUE;
    rightTmpArray[right - middle] = Integer.MAX_VALUE;

    int i = 0,j = 0;
    for (int k = left; k <= right; k++) {
      if (leftTmpArray[i] < rightTmpArray[j]) {
        A[k] = leftTmpArray[i];
        i++;
      }
      else {
        A[k] = rightTmpArray[j];
        j++;
      }
    }
  }

  public static void beginRecurse(int[] Array,int left,int right) {
    if (right > left) {
      int m = (left + right) / 2;
      // recursively call on smaller sections of array and build back up
      beginRecurse(Array,left,m);
      beginRecurse(Array,m + 1,right);
      merge1(Array,left,m,right);
    }
  }

  public static void printArr(int []array) {
    for (int i = 0; i < array.length; i++) {
      System.out.print(array[i] + " ");
    }
  }

  public static void main(String[] args) {
    System.out.println("**** merge sort ****");
    int[] arr = {9,7,25,0,14,71,5,-18,20};

    System.out.println("before sort: " + Arrays.toString(arr));
    MergeSort ms = new MergeSort();
    ms.beginRecurse(arr,0,arr.length - 1);
    System.out.println("after sort: " + Arrays.toString(arr));
  }
}
